src/wallet/message_store.cpp

```diff
@@ -0,0 +1,1389 @@
+  // produce errors if that wallet contains additional members that have nothing to do with those arriving
+  // in "member_config".
+  std::vector<coalition_member> members;
+  unpack_member_config(state, member_config, members);
+  
+  uint32_t new_index = 1;
+  for (size_t i = 0; i < members.size(); ++i)
+  {
+    const coalition_member &m = members[i];
+    uint32_t index;
+    uint32_t take_index;
+    bool take = true;
+    bool found = get_member_index_by_monero_address(m.monero_address, index);
+    if (found)
+    {
+      // Redefine existing (probably "me", under usual circumstances)
+      take_index = index;
+    }
+    else
+    {
+      // Add new; neglect that we may erroneously overwrite already defined members
+      take_index = new_index;
+      new_index++;
+    }
+    if (take)
+    {
+      coalition_member &modify = m_members[take_index];
+      modify.label = m.label;  // ALWAYS set label, see comments above
+      if (!modify.me)
+      {
+        modify.transport_address = m.transport_address;
+        modify.monero_address_known = m.monero_address_known;
+        if (m.monero_address_known)
+        {
+          modify.monero_address = m.monero_address;
+        }
+      }
+    }
+  }
+  save(state);
+}
+
+void message_store::start_auto_config(const multisig_wallet_state &state)
+{
+  for (size_t i = 0; i < m_members.size(); ++i)
+  {
+    coalition_member &m = m_members[i];
+    if (!m.me)
+    {
+      setup_member_for_auto_config(i, create_auto_config_token(), true);
+    }
+    m.auto_config_running = true;
+  }
+  save(state);
+}
+
+// Convert auto-config token string from hex to binary and verify the checksum
+bool message_store::is_auto_config_token_valid(const std::string &auto_config_token)
+{
+  uint32_t good_string_length = (AUTO_CONFIG_TOKEN_BYTES + 1) * 2;
+  if (auto_config_token.length() != good_string_length)
+  {
+    return false;
+  }
+  std::string token_bytes;
+  bool success = epee::string_tools::parse_hexstr_to_binbuff(auto_config_token, token_bytes);
+  if (!success)
+  {
+    return false;
+  }
+  uint32_t checksum = 0;
+  for (size_t i = 0; i < AUTO_CONFIG_TOKEN_BYTES; ++i) {
+    checksum += static_cast<uint8_t>(token_bytes[i]);
+  }
+  success = static_cast<uint8_t>(token_bytes[AUTO_CONFIG_TOKEN_BYTES]) == (checksum % 256);
+  if (!success)
+  {
+    return false;
+  }
+  return true;
+}
+
+// Create a new, random 8-hex digit plus 2 checksum digit auto-config token
+std::string message_store::create_auto_config_token()
+{
+  unsigned char random[AUTO_CONFIG_TOKEN_BYTES];
+  crypto::rand(AUTO_CONFIG_TOKEN_BYTES, random);
+  std::string token_bytes;;
+  token_bytes.append((char *)random, AUTO_CONFIG_TOKEN_BYTES);
+
+  // Add a checksum because technically ANY four bytes are a valid token, and without a checksum we would send
+  // auto-config messages "to nowhere" after the slightest typo without knowing it
```

Comment:
The auto-config mechanism works like that: You receive from me an *auto-config token* e.g. by SMS, phone, or other "reasonably secure" channel; you type that in; the MMS uses the token as seed to derive a single-use Bitmessage address and sends a special auto-config data message to that address, containing your Monero address and your "real" Bitmessage address that you want to use for transacting with me. Because my wallet generated the token it knows on which single-use Bitmessage your data will arrive. Note that auto-config data is encrypted, with a key pair that was also derived from the auto-config token.

If you make a mistake when typing in the token, the MMS would send your auto-config data to a completely different Bitmessage address. Result: My wallet does not receive anything.

I think a primitive checksum should already do the job.

Reply:
Thanks, that makes sense. You have about half a percent chance of the checksum agreeing on a typo with that system. But things like typing "45" instead of "54" will go unnoticed, and that's a common typo type. Will you consider a cn_hash_fast + prefix ?

Reply:
Do you happen to know if the *last* (or *first*?) byte of a CN fast hash **alone** can serve as a good "checksum" that also catches entry errors like switching digits with a good chance? I think it's important to keep those auto-config tokens short, for good usability, hence the question to take only one byte from the whole hash.

With prefix, do you mean something like prepending "mms", so instead of simply `1abc456f33` we would have `mms1abc456f33` as tokens?

Reply:
Yes, any byte of cn_fast_hash would be better.

Reply:
I like the idea with the prefix, so that if you ever find something floating around you will know it's a MMS auto-config token. I reworked the whole thing to use `cn_fast_hash` and "mms" prefix.

When entering tokens I built in tolerance i.e. it works with or without the prefix and with uppercase or lowercase characters for prefix and hex digits. I want people to get those tokens accross even if they dictate them to each other over the phone :)